sql/grant_tables/privilege_set.go

```diff
@@ -0,0 +1,614 @@
+func (ps PrivilegeSetDatabase) Has(privileges ...PrivilegeType) bool {
+	for _, priv := range privileges {
+		if _, ok := ps.privs[priv]; !ok {
+			return false
+		}
+	}
+	return true
+}
+
+// Count returns the number of database privileges.
+func (ps PrivilegeSetDatabase) Count() int {
+	return len(ps.privs)
+}
+
+// Table returns the set of privileges for the given table. Returns an empty set if the table does not exist.
+func (ps PrivilegeSetDatabase) Table(tblName string) PrivilegeSetTable {
+	tblSet, ok := ps.tables[strings.ToLower(tblName)]
+	if ok {
+		return tblSet
+	}
+	return PrivilegeSetTable{name: tblName}
+}
+
+// GetTables returns all tables.
+func (ps PrivilegeSetDatabase) GetTables() []PrivilegeSetTable {
+	tblSets := make([]PrivilegeSetTable, len(ps.tables))
+	i := 0
+	for _, tblSet := range ps.tables {
+		tblSets[i] = tblSet
+		i++
+	}
+	return tblSets
+}
+
+// Equals returns whether the given set of privileges is equivalent to the calling set.
+func (ps PrivilegeSetDatabase) Equals(otherPs PrivilegeSetDatabase) bool {
+	if len(ps.privs) != len(otherPs.privs) ||
+		len(ps.tables) != len(otherPs.tables) {
+		return false
+	}
+	for priv := range ps.privs {
+		if _, ok := otherPs.privs[priv]; !ok {
+			return false
+		}
+	}
+	for tblName, tblSet := range ps.tables {
+		if !tblSet.Equals(otherPs.tables[tblName]) {
+			return false
+		}
+	}
+	return true
+}
+
+// ToSlice returns all of the database privileges contained as a slice.
+func (ps PrivilegeSetDatabase) ToSlice() []PrivilegeType {
+	privs := make([]PrivilegeType, len(ps.privs))
+	i := 0
+	for priv := range ps.privs {
+		privs[i] = priv
+		i++
+	}
+	return privs
+}
+
+// ToSortedSlice returns all of the database privileges contained as a slice, sorted by their internal ID.
+func (ps PrivilegeSetDatabase) ToSortedSlice() []PrivilegeType {
+	privs := ps.ToSlice()
+	sort.Slice(privs, func(i, j int) bool {
+		return privs[i] < privs[j]
+	})
+	return privs
+}
+
+// getUseableTbl is used internally to either retrieve an existing table, or create a new one that is returned.
+func (ps PrivilegeSetDatabase) getUseableTbl(tblName string) PrivilegeSetTable {
+	lowerTblName := strings.ToLower(tblName)
+	tblSet, ok := ps.tables[lowerTblName]
+	if !ok {
+		tblSet = PrivilegeSetTable{
+			name:    tblName,
+			privs:   make(map[PrivilegeType]struct{}),
+			columns: make(map[string]PrivilegeSetColumn),
+		}
+		ps.tables[lowerTblName] = tblSet
+	}
+	return tblSet
+}
+
+// merge merges the given set of privileges to the calling set of privileges.
+func (ps PrivilegeSetDatabase) merge(otherPs PrivilegeSetDatabase) {
+	for priv := range otherPs.privs {
+		ps.privs[priv] = struct{}{}
+	}
+	for _, otherTblSet := range otherPs.tables {
+		ps.getUseableTbl(otherTblSet.name).merge(otherTblSet)
+	}
+}
+
+// clear removes all database privileges.
+func (ps PrivilegeSetDatabase) clear() {
+	for priv := range ps.privs {
+		delete(ps.privs, priv)
```

Comment:
Might just make a new map at this piont

Reply:
This isn't a pointer method but a struct one. Without going into the details of exactly how it's beneficial for later, all of the `PrivilegeSet___` types are returned as structs so that their zero values behave the same as their fully-initialized counterparts. I played around with making valid nil pointer types, but wasn't a fan of the implementation. As a consequence, I can't replace the map, only modify it from here.